File: packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/VideoCaptureTest.java
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.camerax;

import static org.junit.Assert.assertEquals;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import androidx.camera.video.VideoCapture;
import androidx.camera.video.VideoOutput;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.MockedStatic;
import org.mockito.Mockito;
import org.mockito.stubbing.Answer;
import org.robolectric.RobolectricTestRunner;

@RunWith(RobolectricTestRunner.class)
public class VideoCaptureTest {
  @SuppressWarnings({"unchecked", "rawtypes"})
  @Test
  public void withOutput_createsVideoCaptureWithVideoOutput() {
    final PigeonApiVideoCapture api = new TestProxyApiRegistrar().getPigeonApiVideoCapture();

    final VideoCapture<VideoOutput> instance = mock(VideoCapture.class);
    final VideoOutput videoOutput = mock(VideoOutput.class);

    try (MockedStatic<VideoCapture> mockedCamera2CameraInfo =
        Mockito.mockStatic(VideoCapture.class)) {
      mockedCamera2CameraInfo
          .when(() -> VideoCapture.withOutput(videoOutput))
          .thenAnswer((Answer<VideoCapture>) invocation -> instance);

      assertEquals(api.withOutput(videoOutput), instance);
    }
  }

  @SuppressWarnings("unchecked")
  @Test
  public void getOutput_returnsAssociatedRecorder() {
    final PigeonApiVideoCapture api = new TestProxyApiRegistrar().getPigeonApiVideoCapture();

    final VideoCapture<VideoOutput> instance = mock(VideoCapture.class);
    final VideoOutput value = mock(VideoOutput.class);
    when(instance.getOutput()).thenReturn(value);

    assertEquals(value, api.getOutput(instance));
  }

  @SuppressWarnings("unchecked")
  @Test
  public void setTargetRotation_makesCallToSetTargetRotation() {
    final PigeonApiVideoCapture api = new TestProxyApiRegistrar().getPigeonApiVideoCapture();

    final VideoCapture<VideoOutput> instance = mock(VideoCapture.class);
    final long rotation = 0;
    api.setTargetRotation(instance, rotation);

    verify(instance).setTargetRotation((int) rotation);
  }
}

File: packages/camera/camera_android_camerax/test/video_capture_test.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:camera_android_camerax/src/camerax_library.g.dart';
import 'package:camera_android_camerax/src/instance_manager.dart';
import 'package:camera_android_camerax/src/recorder.dart';
import 'package:camera_android_camerax/src/surface.dart';
import 'package:camera_android_camerax/src/video_capture.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:mockito/annotations.dart';
import 'package:mockito/mockito.dart';

import 'test_camerax_library.g.dart';
import 'video_capture_test.mocks.dart';

@GenerateMocks(
    <Type>[TestVideoCaptureHostApi, TestInstanceManagerHostApi, Recorder])
void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  // Mocks the call to clear the native InstanceManager.
  TestInstanceManagerHostApi.setup(MockTestInstanceManagerHostApi());

  test('withOutput calls the Java side and returns correct video capture',
      () async {
    final MockTestVideoCaptureHostApi mockApi = MockTestVideoCaptureHostApi();
    TestVideoCaptureHostApi.setup(mockApi);

    final InstanceManager instanceManager = InstanceManager(
      onWeakReferenceRemoved: (_) {},
    );

    final Recorder mockRecorder = MockRecorder();
    const int mockRecorderId = 2;
    instanceManager.addHostCreatedInstance(mockRecorder, mockRecorderId,
        onCopy: (_) => MockRecorder());

    final VideoCapture videoCapture =
        VideoCapture.detached(instanceManager: instanceManager);
    const int videoCaptureId = 3;
    instanceManager.addHostCreatedInstance(videoCapture, videoCaptureId,
        onCopy: (_) => VideoCapture.detached(instanceManager: instanceManager));

    when(mockApi.withOutput(mockRecorderId)).thenReturn(videoCaptureId);

    expect(
        await VideoCapture.withOutput(mockRecorder,
            instanceManager: instanceManager),
        videoCapture);
    verify(mockApi.withOutput(mockRecorderId));
  });

  test(
      'setTargetRotation makes call to set target rotation for VideoCapture instance',
      () async {
    final MockTestVideoCaptureHostApi mockApi = MockTestVideoCaptureHostApi();
    TestVideoCaptureHostApi.setup(mockApi);

    final InstanceManager instanceManager = InstanceManager(
      onWeakReferenceRemoved: (_) {},
    );
    const int targetRotation = Surface.rotation180;
    final VideoCapture videoCapture = VideoCapture.detached(
      instanceManager: instanceManager,
    );
    instanceManager.addHostCreatedInstance(
      videoCapture,
      0,
      onCopy: (_) => VideoCapture.detached(instanceManager: instanceManager),
    );

    await videoCapture.setTargetRotation(targetRotation);

    verify(mockApi.setTargetRotation(
        instanceManager.getIdentifier(videoCapture), targetRotation));
  });

  test('getOutput calls the Java side and returns correct Recorder', () async {
    final MockTestVideoCaptureHostApi mockApi = MockTestVideoCaptureHostApi();
    TestVideoCaptureHostApi.setup(mockApi);
    final InstanceManager instanceManager = InstanceManager(
      onWeakReferenceRemoved: (_) {},
    );

    final VideoCapture videoCapture =
        VideoCapture.detached(instanceManager: instanceManager);
    const int videoCaptureId = 2;
    instanceManager.addHostCreatedInstance(videoCapture, videoCaptureId,
        onCopy: (_) => VideoCapture.detached(instanceManager: instanceManager));

    final Recorder mockRecorder = MockRecorder();
    const int mockRecorderId = 3;
    instanceManager.addHostCreatedInstance(mockRecorder, mockRecorderId,
        onCopy: (_) => Recorder.detached(instanceManager: instanceManager));

    when(mockApi.getOutput(videoCaptureId)).thenReturn(mockRecorderId);
    expect(await videoCapture.getOutput(), mockRecorder);
    verify(mockApi.getOutput(videoCaptureId));
  });

  test('flutterApiCreateTest', () async {
    final InstanceManager instanceManager = InstanceManager(
      onWeakReferenceRemoved: (_) {},
    );

    final VideoCaptureFlutterApi flutterApi = VideoCaptureFlutterApiImpl(
      instanceManager: instanceManager,
    );

    flutterApi.create(0);

    expect(
        instanceManager.getInstanceWithWeakReference(0), isA<VideoCapture>());
  });
}
